examples/js/nodes/materials/NodeMaterial.js

```diff
@@ -18,6 +18,22 @@ function NodeMaterial( vertex, fragment ) {
 
 	this.updaters = [];
 
+	// onBeforeCompile can't be in the prototype because onBeforeCompile.toString varies per material
```

Comment:
Where does `onBeforeCompile.toString` get called?

Reply:
It is used to create a hash modification in onBeforeCompile. More details:
https://github.com/mrdoob/three.js/pull/14442
https://github.com/mrdoob/three.js/pull/16668

Reply:
https://github.com/mrdoob/three.js/pull/15674

Reply:
I don't see anything calling `toString()` in these PRs?

Reply:
I'm confused,threejs is a new unfamiliar technology for me ,how to learning it ?

Reply:
In 15674 have a commit calling toString
https://github.com/mrdoob/three.js/pull/15674/commits/07236bb765fd2d2231196ca89125b335144241e1
The description of the problem:
https://github.com/mrdoob/three.js/pull/15674#issuecomment-459417579

I think that this approach will be improved in the future...
https://github.com/mrdoob/three.js/blob/7fe9a41cb6930d8600305e8ee85830b011e461bb/src/renderers/webgl/WebGLPrograms.js#L249